File: deskshare/app/src/main/scala/org/bigbluebutton/deskshare/server/svc1/BlockFactory.scala
/**
* BigBlueButton open source conferencing system - http://www.bigbluebutton.org/
* 
* Copyright (c) 2012 BigBlueButton Inc. and by respective authors (see below).
*
* This program is free software; you can redistribute it and/or modify it under the
* terms of the GNU Lesser General Public License as published by the Free Software
* Foundation; either version 3.0 of the License, or (at your option) any later
* version.
* 
* BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
* WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
* PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
*
*/
package org.bigbluebutton.deskshare.server.svc1

trait BlockFactory {
  
	def getNumberOfColumns(screenDim: Dimension , blockDim: Dimension ): Int = {
    	var columns: Int = screenDim.width / blockDim.width
    	if (hasPartialColumnBlock(screenDim, blockDim)) {
    		columns += 1
    	}
     
    	return columns;
    }
    
    private def hasPartialColumnBlock(screenDim: Dimension , blockDim: Dimension): Boolean = {
    	return (screenDim.width % blockDim.width) != 0
    }
    
    def getNumberOfRows(screenDim: Dimension, blockDim: Dimension ): Int = {
    	var rows: Int = screenDim.height / blockDim.height
    	if (hasPartialRowBlock(screenDim, blockDim)) {
    		rows += 1
    	}
    	return rows
    }
    
    private def hasPartialRowBlock(screenDim: Dimension, blockDim: Dimension): Boolean = {
    	return (screenDim.height % blockDim.height) != 0;
    }
  
    def createBlock(screenDim: Dimension, blockDim: Dimension, position: Int): Block = {
    	val numRows: Int = getNumberOfRows(screenDim, blockDim)
    	val numColumns: Int = getNumberOfColumns(screenDim, blockDim)
    	
    	val col: Int = computeColumn(position, numColumns)
    	val row: Int = computeRow(position, numRows, numColumns)
		val w: Int = computeTileWidth(col, screenDim, blockDim)
		val h: Int = computeTileHeight(row, screenDim, blockDim)	
		
		return new Block(new Dimension(w, h), position) 
    }
       
    private def computeRow(position: Int, numRows: Int, numColumns: Int): Int = {
    	return -(position - (numRows * numColumns)) / numColumns
    }
    
    private def computeColumn(position: Int, numColumns: Int): Int = {
		return (position - 1) % numColumns    	
    }
    
    private def computeTileWidth(col: Int, screenDim: Dimension, blockDim: Dimension): Int = {
    	val numColumns: Int = getNumberOfColumns(screenDim, blockDim)
    	if (isLastColumnTile(col, numColumns)) {
    		if (hasPartialColumnBlock(screenDim, blockDim)) {
    			return partialTileWidth(screenDim, blockDim)
    		}
    	}
    	return blockDim.width
    }
    
    private def partialTileWidth(screenDim: Dimension, blockDim: Dimension): Int = {
    	return screenDim.width % blockDim.width
    }
    
    private def computeTileHeight(row: Int, screenDim: Dimension, blockDim: Dimension): Int = {
    	if (isTopRowTile(row)) {
    		if (hasPartialRowBlock(screenDim, blockDim)) {
    			return partialTileHeight(screenDim, blockDim)
    		}
    	}
    	return blockDim.width
    }
    
    private def partialTileHeight(screenDim: Dimension, blockDim: Dimension): Int = {
    	return screenDim.height % blockDim.height
    }

    private def  isLastColumnTile(col: Int, numColumns: Int): Boolean = {
    	return ((col+1) % numColumns) == 0;
    }
    
    private def  isTopRowTile(row: Int): Boolean = {
    	return (row == 0);
    }
}

File: deskshare/applet/src/main/java/org/bigbluebutton/deskshare/client/blocks/BlockFactory.java
/**
* BigBlueButton open source conferencing system - http://www.bigbluebutton.org/
* 
* Copyright (c) 2012 BigBlueButton Inc. and by respective authors (see below).
*
* This program is free software; you can redistribute it and/or modify it under the
* terms of the GNU Lesser General Public License as published by the Free Software
* Foundation; either version 3.0 of the License, or (at your option) any later
* version.
* 
* BigBlueButton is distributed in the hope that it will be useful, but WITHOUT ANY
* WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
* PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License along
* with BigBlueButton; if not, see <http://www.gnu.org/licenses/>.
*
*/
package org.bigbluebutton.deskshare.client.blocks;

import java.awt.Point;

import org.bigbluebutton.deskshare.common.Dimension;

class BlockFactory {
    private int numColumns;
    private int numRows;
    private Dimension screenDim;
    private Dimension tileDim;
        
    public BlockFactory(Dimension screen, Dimension tile) {
        this.screenDim = screen;
        this.tileDim = tile;
        
        numColumns = computeNumberOfColumnTiles();
        numRows = computeNumberOfRowTiles();
    }
    
    private int computeNumberOfColumnTiles() {
    	int columns = screenDim.getWidth() / tileDim.getWidth();
    	if (hasPartialColumnTile()) {
    		columns += 1;
    	}
    	return columns;
    }
    
    private boolean hasPartialColumnTile() {
    	return (screenDim.getWidth() % tileDim.getWidth()) != 0;
    }
    
    private int computeNumberOfRowTiles() {
    	int rows = screenDim.getHeight() / tileDim.getHeight();
    	if (hasPartialRowTile()) {
    		rows += 1;
    	}
    	return rows;
    }
    
    private boolean hasPartialRowTile() {
    	return (screenDim.getHeight() % tileDim.getHeight()) != 0;
    }
  
    Block createBlock(int position, boolean useSVC2) {
    	int col = computeColumn(position);
    	int row = computeRow(position);
		int w = computeTileWidth(col);
		int h = computeTileHeight(row);		
		int x = computeTileXLocation(col);
		int y = computeTileYLocation(row);
		int pos = computeTilePosition(row, col);
		
		Block t = new Block(new Dimension(w, h), pos, new Point(x,y), useSVC2);

		return t;
    }
    
    private int computeTilePosition(int row, int col) {
    	return (((numRows - (row+1)) * numColumns) + (col + 1));
    }
    
    private int computeTileXLocation(int col) {
    	return col * tileDim.getWidth();
    }
    
    private int computeTileYLocation(int row) {
    	if (isTopRowTile(row)) return 0;
    	return screenDim.getHeight() - ((numRows - row) * tileDim.getHeight());
    }
    
    Point positionToIndex(int position) {    
    	return new Point(computeRow(position), computeColumn(position));
    }
    
    int indexToPosition(int row, int col) {
    	return computeTilePosition(row, col);
    }
    
    private int computeRow(int position) {
    	return -(position - (getRowCount() * getColumnCount())) / getColumnCount();
    }
    
    private int computeColumn(int position) {
		return (position - 1) % getColumnCount();    	
    }
    
    private int computeTileWidth(int col) {
    	if (isLastColumnTile(col)) {
    		if (hasPartialColumnTile()) {
    			return partialTileWidth();
    		}
    	}
    	return tileDim.getWidth();
    }
    
    private int partialTileWidth() {
    	return screenDim.getWidth() % tileDim.getWidth();
    }
    
    private int computeTileHeight(int row) {
    	if (isTopRowTile(row)) {
    		if (hasPartialRowTile()) {
    			return partialTileHeight();
    		}
    	}
    	return tileDim.getWidth();
    }
    
    private int partialTileHeight() {
    	return screenDim.getHeight() % tileDim.getHeight();
    }

    private boolean isLastColumnTile(int col) {
    	return ((col+1) % numColumns) == 0;
    }
    
    private boolean isTopRowTile(int row) {
    	return (row == 0);
    }

    
    int getRowCount()
    {
        return numRows;
    }
    
    int getColumnCount()
    {
        return numColumns;
    }
}
